firedrake/output.py

```diff
@@ -651,3 +756,366 @@ def write(self, *functions, **kwargs):
+        self.set_attr(path, PREFIX + "_ufl_element_family", tV.ufl_element().family())
+        self.set_attr(path, PREFIX + "_ufl_element_degree", tV.ufl_element().degree())
+
+    def save_function(self, f):
+        V = f.function_space()
+        # Save function space
+        self.save_function_space(V)
+        if type(V.topological) is impl.MixedFunctionSpace:
+            base_path = os.path.join("/topologies",
+                                     PREFIX + "_mixed_function_spaces",
+                                     V.name,
+                                     PREFIX + "_functions",
+                                     f.name())
+            self.h5pyfile.require_group(base_path)
+            for i, fsub in enumerate(f.split()):
+                path = os.path.join(base_path, str(i))
+                self.h5pyfile.require_group(path)
+                self.set_attr(path, PREFIX + "_function", fsub.name())
+                self.save_function(fsub)
+            return
+        # Save function
+        tf = f.topological
+        tV = tf.function_space()
+        dm = tV._mesh.topology_dm
+        sectiondm = get_sectiondm_for_checkpointing(tV)
+        path = os.path.join("/topologies",
+                            dm.getName(),
+                            "dms",
+                            sectiondm.getName(),
+                            "vecs",
+                            tf.name())
+        self.h5pyfile.require_group(path)
+        self.set_attr(path, PREFIX + "_function_space", f.function_space().name)
+        # -- Embed if necessary
+        element = V.ufl_element()
+        _element, method = get_embedding_element(element)
+        self.set_attr(path, PREFIX + "_embedded", _element != element)
+        if _element != element:
+            _V = FunctionSpace(V.mesh(), _element, name=PREFIX_EMBEDDED + V.name)
+            _f = Function(_V, name=PREFIX_EMBEDDED + f.name())
+            if method == "project":
+                getattr(_f, method)(f, solver_parameters={"ksp_rtol": 1.e-16})
+            else:
+                getattr(_f, method)(f)
+            self.save_function(_f)
+            return
+        vec = tf.dat._vec
+        vec.setName(tf.name())
+        dm.globalVectorView(self.viewer, sectiondm, vec)
+
+    def load_mesh(self, name=DEFAULT_MESH_NAME, distribution_parameters={}):
+        dist_key = hash(frozenset(distribution_parameters.items()))
+        mesh_key = (name, dist_key)
+        if mesh_key in self._meshes:
+            return self._meshes[mesh_key]["mesh"]
+        plex, sfXB = _dmload_plex(self.viewer, name)
+        # These labels are distribution dependent.
+        # We should be able to save/load labels selectively.
+        plex.removeLabel("pyop2_core")
+        plex.removeLabel("pyop2_owned")
+        plex.removeLabel("pyop2_ghost")
+        mesh = Mesh(plex, name=name, distribution_parameters=distribution_parameters)
```

Comment:
Is the default to do redistribution (for better load balance)? If not, it should be.